### PR TITLE
Use Bytes instead of &[u8] to reduce memory copying

### PR DESCRIPTION
**Pull request by YHM404**
When using &[u8], it's hard to avoid memory copying due to lifetime issues. Replacing &[u8] with Bytes can effectively reduce memory copying.

### CODE REVIEW

**Comment by hubcio** [comment]
for me server panics:
```
thread 'tokio-runtime-worker' panicked at sdk/src/command.rs:261:47:
range end index 4 out of range for slice of length 0
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

when running ` target/release/iggy-bench -v send tcp`

**Comment by hubcio** [comment]
please run bench and attach before/after to the PR

**Comment by YHM404** [comment]
> please run bench and attach before/after to the PR

This PR is still in progress. 
There are strange performance issues.

**Comment by YHM404** [comment]
**Before**
Send TCP
```
2024-01-30T17:40:44.088953Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 1120.89 MB/s, 1120894 messages/s, average latency: 8.92 ms, average  throughput: 112.09 MB/s, total duration: 8.92 s
```

Poll TCP
```
2024-01-30T17:41:48.839996Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 2108.67 MB/s, 2017868 messages/s, average latency: 4.85 ms, average  throughput: 210.87 MB/s, total duration: 4.96 s
```

Send && poll TCP
```
2024-01-30T17:40:10.334675Z  INFO iggy_bench::benchmark_runner: Producer results: total throughput: 1822.24 MB/s, 1822238 messages/s, average latency: 5.49 ms, average throughput: 91.11 MB/s, total duration: 5.49 s, Consumer results: total throughput: 1882.17 MB/s, 1801123 messages/s, average latency: 5.36 ms, average  throughput: 94.11 MB/s, total duration: 5.55 s
2024-01-30T17:40:10.334678Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 1852.38 MB/s, 1811619 messages/s, average latency: 10.85 ms, average  throughput: 92.62 MB/s, total duration: 11.04 s
```

**After**
Send TCP
```
2024-01-30T17:32:46.093686Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 1078.68 MB/s, 1078682 messages/s, average latency: 9.27 ms, average  throughput: 107.87 MB/s, total duration: 9.27 s
```

poll TCP
```
2024-01-30T17:33:20.772766Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 2026.42 MB/s, 1939158 messages/s, average latency: 5.05 ms, average  throughput: 202.64 MB/s, total duration: 5.16 s
```

Send && poll TCP
```
2024-01-30T17:35:47.037283Z  INFO iggy_bench::benchmark_runner: Producer results: total throughput: 2208.14 MB/s, 2208138 messages/s, average latency: 4.53 ms, average throughput: 110.41 MB/s, total duration: 4.53 s, Consumer results: total throughput: 2232.02 MB/s, 2135903 messages/s, average latency: 4.57 ms, average  throughput: 111.60 MB/s, total duration: 4.68 s
2024-01-30T17:35:47.037285Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 2220.28 MB/s, 2171420 messages/s, average latency: 9.09 ms, average  throughput: 111.01 MB/s, total duration: 9.21 s
```

**Comment by YHM404** [comment]
> **Before** Send TCP
> 
> ```
> 2024-01-30T17:40:44.088953Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 1120.89 MB/s, 1120894 messages/s, average latency: 8.92 ms, average  throughput: 112.09 MB/s, total duration: 8.92 s
> ```
> 
> Poll TCP
> 
> ```
> 2024-01-30T17:41:48.839996Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 2108.67 MB/s, 2017868 messages/s, average latency: 4.85 ms, average  throughput: 210.87 MB/s, total duration: 4.96 s
> ```
> 
> Send && poll TCP
> 
> ```
> 2024-01-30T17:40:10.334675Z  INFO iggy_bench::benchmark_runner: Producer results: total throughput: 1822.24 MB/s, 1822238 messages/s, average latency: 5.49 ms, average throughput: 91.11 MB/s, total duration: 5.49 s, Consumer results: total throughput: 1882.17 MB/s, 1801123 messages/s, average latency: 5.36 ms, average  throughput: 94.11 MB/s, total duration: 5.55 s
> 2024-01-30T17:40:10.334678Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 1852.38 MB/s, 1811619 messages/s, average latency: 10.85 ms, average  throughput: 92.62 MB/s, total duration: 11.04 s
> ```
> 
> **After** Send TCP
> 
> ```
> 2024-01-30T17:32:46.093686Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 1078.68 MB/s, 1078682 messages/s, average latency: 9.27 ms, average  throughput: 107.87 MB/s, total duration: 9.27 s
> ```
> 
> poll TCP
> 
> ```
> 2024-01-30T17:33:20.772766Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 2026.42 MB/s, 1939158 messages/s, average latency: 5.05 ms, average  throughput: 202.64 MB/s, total duration: 5.16 s
> ```
> 
> Send && poll TCP
> 
> ```
> 2024-01-30T17:35:47.037283Z  INFO iggy_bench::benchmark_runner: Producer results: total throughput: 2208.14 MB/s, 2208138 messages/s, average latency: 4.53 ms, average throughput: 110.41 MB/s, total duration: 4.53 s, Consumer results: total throughput: 2232.02 MB/s, 2135903 messages/s, average latency: 4.57 ms, average  throughput: 111.60 MB/s, total duration: 4.68 s
> 2024-01-30T17:35:47.037285Z  INFO iggy_bench::benchmark_runner: Results: total throughput: 2220.28 MB/s, 2171420 messages/s, average latency: 9.09 ms, average  throughput: 111.01 MB/s, total duration: 9.21 s
> ```

cc @hubcio

File: sdk/src/binary/mapper.rs [hunk 1]
```diff
@@ -104,13 +104,13 @@ pub fn map_consumer_offset(payload: &[u8]) -> Result<ConsumerOffsetInfo, IggyErr
     })
 }
 
-pub fn map_user(payload: &[u8]) -> Result<UserInfoDetails, IggyError> {
-    let (user, position) = map_to_user_info(payload, 0)?;
+pub fn map_user(payload: Bytes) -> Result<UserInfoDetails, IggyError> {
+    let (user, position) = map_to_user_info(payload.clone(), 0)?;
```
Comment (hubcio) [inline sdk/src/binary/mapper.rs:108]: why clone and not pass payload as reference?

File: sdk/src/binary/mapper.rs [hunk 2]
```diff
@@ -154,30 +154,30 @@ pub fn map_personal_access_tokens(
     let length = payload.len();
     let mut position = 0;
     while position < length {
-        let (personal_access_token, read_bytes) = map_to_pat_info(payload, position)?;
+        let (personal_access_token, read_bytes) = map_to_pat_info(payload.clone(), position)?;
```
Comment (hubcio) [inline sdk/src/binary/mapper.rs:157]: same as above comment

File: sdk/src/consumer_groups/delete_consumer_group.rs [hunk 3]
```diff
@@ -33,30 +34,30 @@ impl Validatable<IggyError> for DeleteConsumerGroup {
...
 
-    fn from_bytes(bytes: &[u8]) -> Result<DeleteConsumerGroup, IggyError> {
+    fn from_bytes(bytes: Bytes) -> Result<DeleteConsumerGroup, IggyError> {
         if bytes.len() < 9 {
             return Err(IggyError::InvalidCommand);
         }
 
         let mut position = 0;
-        let stream_id = Identifier::from_bytes(bytes)?;
+        let stream_id = Identifier::from_bytes(bytes.clone())?;
```
Comment (hubcio) [inline sdk/src/consumer_groups/delete_consumer_group.rs:56]: like above, consider changing signature of from_bytes to accept ref and perhaps try to pass a reference

**Review by hubcio — COMMENTED** [review]
overall, very nice!!!!

File: sdk/src/consumer_groups/delete_consumer_group.rs (hunk 3 above)
Comment (YHM404) [reply sdk/src/consumer_groups/delete_consumer_group.rs:56]: https://docs.rs/bytes/1.5.0/bytes/struct.Bytes.html

ref：`They are cheaply cloneable and thereby shareable between an unlimited amount of components, for example by modifying a reference count.`

https://docs.rs/bytes/1.5.0/bytes/struct.Bytes.html#sharing

Bytes clone and passing &Bytes incur almost the same cost, it doesn't actually copy the underlying data, just increases the reference count.

**Review by spetz — APPROVED** [review]
Works great, thank you for this improvement!

**Pull request merged**